contract/src/stage1/delete.rs

```diff
@@ -1,40 +1,223 @@
+
+            // Loop through and remove keys
+            for key in &keys_to_delete {
+                // Unlink key to drop ID
+                self.drop_id_for_pk.remove(key);
+                // Attempt to remove the public key. panic if it didn't exist
+                require!(drop.pks.remove(key) == true, "public key must be in drop");
+            }
+
+            /*
+                Refund amount consists of:
+                - Storage for each key
+                - Access key allowance
+                - Access key storage
+                - Balance for linkdrop
+                
+                Optional:
+                - FC deposit
+                - storage for longest token ID
+                - FT storage registration cost
+            */ 
+            let fc_cost = drop.fc_data.as_ref().map(|data| data.deposit.0).unwrap_or(0);
+            let ft_cost = drop.ft_data.as_ref().map(|data| data.ft_storage.unwrap().0).unwrap_or(0);
+            let nft_cost = drop.nft_data.as_ref().map(|data| data.storage_for_longest * env::storage_byte_cost()).unwrap_or(0);
+            
+            total_refund_amount = (drop.storage_used_per_key.0 + self.key_fee + ACCESS_KEY_ALLOWANCE + ACCESS_KEY_STORAGE + drop.balance.0 + fc_cost + ft_cost + nft_cost) * len;
+
+            // If the drop has no keys, remove it from the funder. Otherwise, insert it back with the updated keys.
+            if drop.pks.len() == 0 {
+                env::log_str("Drop empty. Removing from funder");
+                self.internal_remove_drop_for_funder(&funder_id, &drop_id);
+            } else {
+                env::log_str(&format!("Drop non empty. Adding back. Len: {}", drop.pks.len()));
+                self.drop_for_id.insert(&drop_id, &drop);
+            }
+        } else {
+            // If no PKs were passed in, attempt to remove 100 keys at a time
+            keys_to_delete = drop.pks.iter().take(100).collect();
+            let len = keys_to_delete.len() as u128;
+            env::log_str(&format!("Removing {} keys from the drop", len));
+            
+            // Loop through and remove keys
+            for key in &keys_to_delete {
+                // Unlink key to drop ID
+                self.drop_id_for_pk.remove(key);
+                // Remove the PK from the unordered set
+                drop.pks.remove(key);
+            }
+
+            /*
+                Refund amount consists of:
+                - Storage for each key
+                - Access key allowance
+                - Access key storage
+                - Balance for linkdrop
+                
+                Optional:
+                - FC deposit
+                - storage for longest token ID
+                - FT storage registration cost
+            */ 
+            let fc_cost = drop.fc_data.as_ref().map(|data| data.deposit.0).unwrap_or(0);
+            let ft_cost = drop.ft_data.as_ref().map(|data| data.ft_storage.unwrap().0).unwrap_or(0);
```

Comment:
double checking: we're counting the funder's deposited storage_balance_of for each key? So for FCs they are getting this amount back on a per key basis?

Reply:
I'm assuming that's what ft_storage is, just want to make sure it's not "just" the storage cost in our contract, but the deposit we were holding to pay for each key's storage_deposit ya?

Reply:
The `ft_storage` is the registration fee on the FT contract which is calculated in stage 1 via a CCC to the FT contract and calling `storage_balance_bounds`

Reply:
Perfect!